### PR TITLE
Fixes #6234 #7947 #7909  Add karaoke effect for text transformation

### LINKED ISSUE #6234: [Request and Bug] Karaoke Effect
1. Kindly add support for word-level Karaoke Effect?
2. When I add the Karaoke effect to a video, the subtitles tend to flicker. Is this normal and can it be fixed?

### PR DESCRIPTION
**Pull request by ivandrofly**
Added classes for karaoke effect text transformation with support for splitting by character or word. Also implemented a strategy pattern to handle the different transformations. Additionally, added UI components to let users choose the type of effect and made adjustments to existing classes for the new features.

- Karaoke: Legacy strategy is preserved (Karaoke effect)
- Karaoke: Is formatting aware (tag existing tag in text will be keep*)
- Karaoke: Has both per word and per char coloring

Note: The new **Typewriter effect class**  also fixes tag issue but it's not yet integrated with the UI.

@niksedk if you could also test on your side and provide feedback.
Unit-tests will be added in upcoming days

### CODE REVIEW

**Comment by ivandrofly** [comment]
For #7909 if user choose option for **per-word karaoke effect** it fits* with what the user has described

File: src/libse/Common/TextEffect/KaraokeEffect.cs [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+            }
+
+            var len = text.Length;
+            for (var i = 0; i < len; i++)
+            {
+                var ch = text[i];
+                if (HtmlUtil.IsStartTagSymbol(ch))
+                {
+                    var closeIdx = text.IndexOf(HtmlUtil.GetClosingPair(ch), i + 1);
+                    if (closeIdx < 0) return i;
```
Comment (niksedk) [inline src/libse/Common/TextEffect/KaraokeEffect.cs:50]: Please always use brackets

File: src/libse/Common/TextEffect/KaraokeEffect.cs [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+                }
+                else return i;
```
Comment (niksedk) [inline src/libse/Common/TextEffect/KaraokeEffect.cs:52]: Please always use brackets

**Comment by niksedk** [comment]
Hi @ivandrofly ,

Looks good :)

Suggestions:
- Please do remove the "legacy" version + remove the "New effect/legacy effect" (perhaps an setting for white space).
- The gap between lines should be zero (was also 1 in the old version), but start/end time should be the same to avoid flickering (or a setting).
- The `tab-stop order` should be fixed (was also broken in old version). 
- Tabbing out of 'End delay in seconds' changes the value.
- 'End delay in seconds' combobox does not work?

**Comment by ivandrofly** [comment]
> * 'End delay in seconds' combobox does not work?

Combobox? You mean the numeric up/down? It does work, try increasing the value and click preview

> * Tabbing out of 'End delay in seconds' changes the value.

Yes, but you have to click Preview button to play/replay the effect. Do you want it to be reactive as soon as the value changed?

**Comment by niksedk** [comment]
Let me know when it's ready :)

**Comment by ivandrofly** [comment]
> Let me know when it's ready :)

Sure, currently on Mac and Unit-test project doesn't play nice with it :D, maybe we need start considering making unit-test project target multi-platform testing framework as LibSE .

**Comment by niksedk** [comment]
> > Let me know when it's ready :)
> 
> Sure, I'm currently on Mac and Unit-test project doesn't play nice with it :D, maybe we need start considering making unit-test project target multi-platform testing framework as LibSE .

So, it's ready besides unit tests? (they can always be added later)

**Comment by ivandrofly** [comment]
> > > Let me know when it's ready :)
> > 
> > 
> > Sure, I'm currently on Mac and Unit-test project doesn't play nice with it :D, maybe we need start considering making unit-test project target multi-platform testing framework as LibSE .
> 
> So, it's ready besides unit tests? (they can always be added later)

yes

**Comment by ivandrofly** [comment]
1/3 Issues was closed, can you close the remaining @niksedk

**Comment by niksedk** [comment]
Is think #7909 is something else...

**Comment by ivandrofly** [comment]
Oh I think I got it now... something like e.g: 

Input: **Foo bar zz**

Output:
\<font..>foo\</font> bar zz
foo \<font..>bar\</font> zz
foo bar \<font..>zz\</font>

Maybe char/second can be used to try to estimated what word will is to be highlighted.

**Comment by niksedk** [comment]
I think this requires something like:

image

**Comment by ivandrofly** [comment]
> I think this requires something like `--highlight_words true` in Faster Whisper:
> 
> image

Okay, Will take a look - haven't played with Whisper yet

**Comment by niksedk** [comment]
> Okay, Will take a look - haven't played with Whisper yet

It's pretty good, and it's getting even better slowly.... the  `--highlight_words true` actually gives a pretty large block...
The timestamps are sometimes a little off. And a word here and there are also often off.

**Comment by M4Mahdi** [comment]
> Oh I think I got it now... something like e.g:
> 
> Input: **Foo bar zz**
> 
> Output: <font..>foo</font> bar zz foo <font..>bar</font> zz foo bar <font..>zz</font>
> 
> The font open tag will be moving instead of staying in fixed location like in karaoke effect.
> 
> Maybe char/second can be used to try to estimated what word will is to be highlighted.
> 
> Ref #7909

You can check this comment too:
https://github.com/SubtitleEdit/subtitleedit/issues/7909#issuecomment-2067876022

**Pull request merged**